genesyscloud/resource_genesyscloud_knowledge_category.go

```diff
@@ -0,0 +1,281 @@
+package genesyscloud
+
+import (
+	"context"
+	"fmt"
+	"log"
+	"strings"
+	"time"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/resource"
+	"github.com/mypurecloud/terraform-provider-genesyscloud/genesyscloud/consistency_checker"
+
+	"github.com/hashicorp/terraform-plugin-sdk/v2/diag"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/schema"
+	"github.com/hashicorp/terraform-plugin-sdk/v2/helper/validation"
+	"github.com/mypurecloud/platform-client-sdk-go/v80/platformclientv2"
+)
+
+var (
+	knowledgeCategory = &schema.Resource{
+		Schema: map[string]*schema.Schema{
+			"name": {
+				Description: "Knowledge base name",
+				Type:        schema.TypeString,
+				Optional:    true,
+				ForceNew:    true,
+			},
+			"description": {
+				Description: "Knowledge base description",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+			"parent_id": {
+				Description: "Knowledge category parent id",
+				Type:        schema.TypeString,
+				Optional:    true,
+			},
+		},
+	}
+)
+
+func getAllKnowledgeCategories(_ context.Context, clientConfig *platformclientv2.Configuration) (ResourceIDMetaMap, diag.Diagnostics) {
+	knowledgeBaseList := make([]platformclientv2.Knowledgebase, 0)
+	resources := make(ResourceIDMetaMap)
+	knowledgeAPI := platformclientv2.NewKnowledgeApiWithConfig(clientConfig)
+
+	for pageNum := 1; ; pageNum++ {
+		const pageSize = 100
+		knowledgeBases, _, getErr := knowledgeAPI.GetKnowledgeKnowledgebases("", "", "", fmt.Sprintf("%v", pageSize), "", "", false, "", "")
+		if getErr != nil {
+			return nil, diag.Errorf("Failed to get page of knowledge bases: %v", getErr)
+		}
+
+		if knowledgeBases.Entities == nil || len(*knowledgeBases.Entities) == 0 {
+			break
+		}
+		for _, knowledgeBase := range *knowledgeBases.Entities {
+			knowledgeBaseList = append(knowledgeBaseList, knowledgeBase)
+		}
+	}
+	for _, knowledgeBase := range knowledgeBaseList {
+		for pageNum := 1; ; pageNum++ {
+			const pageSize = 100
+			knowledgeCategories, _, getErr := knowledgeAPI.GetKnowledgeKnowledgebaseLanguageCategories(*knowledgeBase.Id, *knowledgeBase.CoreLanguage, "", "", "", fmt.Sprintf("%v", pageSize), "")
+			if getErr != nil {
+				return nil, diag.Errorf("Failed to get page of knowledge categories: %v", getErr)
+			}
+
+			if knowledgeCategories.Entities == nil || len(*knowledgeCategories.Entities) == 0 {
+				break
+			}
+
+			for _, knowledgeCategory := range *knowledgeCategories.Entities {
+				id := fmt.Sprintf("%s %s %s", *knowledgeCategory.Id, *knowledgeCategory.KnowledgeBase.Id, *knowledgeCategory.LanguageCode)
+				resources[id] = &ResourceMeta{Name: *knowledgeCategory.Name}
+			}
+		}
+	}
+
+	return resources, nil
+}
+
+func knowledgeCategoryExporter() *ResourceExporter {
+	return &ResourceExporter{
+		GetResourcesFunc: getAllWithPooledClient(getAllKnowledgeCategories),
+		RefAttrs:         map[string]*RefAttrSettings{}, // No references
```

Comment:
This will need to have the `knowledge_base_id` reference

Reply:
The ref is added there, not sure why this isn't showing as Outdated.

Reply:
This ref is added, not sure why it's showing as Outdated.

Reply:
It looks like there's a permission missing for the acceptance tests
<img width="1291" alt="knowledge-base-403" src="https://user-images.githubusercontent.com/80964224/192005902-439c05a7-d3df-4b32-9d82-a12bf6a799b2.png">

There are also a couple tests for existing resources failing.  Going to look into the media retention policy failure

Reply:
Fix the permission issue

Reply:
@carnellj-genesys Fixed. Also the data source test is fixed now.  It searches for published and unpublished results and takes the published one if possible.